oshi-core/src/main/java/oshi/software/common/AbstractOperatingSystem.java

```diff
@@ -170,29 +163,36 @@ protected static void addChildrenToDescendantSet(Collection<OSProcess> allProcs,
      *            as value
      * @param parentPid
      *            The process ID whose children or descendants to return
-     * @param descendantPids
-     *            On input, an empty set, or a set of other children already
-     *            retrieved when using recursion. On output, the children of the
-     *            parent added to the set.
-     * @param recurse
+     * @param allDescendants
      *            If false, only gets immediate children of this process. If true,
      *            gets all descendants.
+     * @return Set of children or descendants of parentPid
      */
-    protected static void addChildrenToDescendantSet(Map<Integer, Integer> parentPidMap, int parentPid,
-            Set<Integer> descendantPids, boolean recurse) {
+    protected static Set<Integer> getChildrenOrDescendants(Map<Integer, Integer> parentPidMap, int parentPid,
+            boolean allDescendants) {
+        Set<Integer> descendantPids = new HashSet();
+        Queue<Integer> queue = new LinkedList();
+        int currPid;
         // Collect this process's children
-        Set<Integer> childPids = parentPidMap.entrySet().stream().filter(e -> e.getValue().equals(parentPid))
-                .map(Entry::getKey).collect(Collectors.toSet());
+        Set<Integer> childPids = getChildren(parentPidMap, parentPid);
         // Add to descendant set
         descendantPids.addAll(childPids);
-        // Recurse
-        if (recurse) {
-            for (int pid : childPids) {
-                if (pid != parentPid && !descendantPids.contains(pid)) {
-                    addChildrenToDescendantSet(parentPidMap, pid, descendantPids, true);
+        // Add all descendents
+        if (allDescendants) {
+            queue.addAll(descendantPids);
+            while (!queue.isEmpty()) {
+                currPid = queue.remove();
+                // Collect current process's children
+                childPids = getChildren(parentPidMap, currPid);
+                for (int pid : childPids) {
+                    if (pid != parentPid && !descendantPids.contains(pid)) {
```

Comment:
Per my earlier comment if you filter out the parent in the `getChildren()` method you can skip it here.